quint/mempool/Consensus.qnt

```diff
@@ -0,0 +1,203 @@
+      state.decided.get(p)
+    }
+
+    pure def proposalOf(state: ConsensusState, p: Process): Set[Tx] = {
+      state.proposed.get(p)
+    }
+
+    pure def getProposed(state: ConsensusState): Set[Tx] = {
+      flatten(mapValues(state.proposed))
+    }
+
+    pure def getDecided(state: ConsensusState): Set[Tx] = {
+      flatten(mapValues(state.decided))
+    }
+
+    pure def isSupersededBy(state0: ConsensusState, state1: ConsensusState): bool = {
+      and {
+        state0.processes() == state1.processes(),
+	state0.processes().forall(
+	  p =>
+	    or {
+	      not(state0.hasDecided(p)) and state1.hasDecided(p),
+	      state0.hasDecided(p) and state1.hasDecided(p)
+	    }
+        )
+      }
+    }
+
+    // 2. preconditions
+
+    pure def mayPropose(state: ConsensusState, p: Process, txs: Set[Tx]): bool = {
+      state.processes().contains(p) and not(state.hasProposed(p)) and txs.size() != 0
+    }
+
+    pure def mayDecide(state: ConsensusState, p: Process): bool = {
+      processes(state).contains(p) and size(state.getProposed()) > 0 and not(state.hasDecided(p))
+    }
+
+    // 3. transitions
+
+    pure def propose(state: ConsensusState, p: Process, txs: Set[Tx]): ConsensusState = {
+      val newProposed = state.proposed.set(p, txs)
+      {proposed: newProposed, ...state}
+    }
+
+    // decide any proposed of the values
+    pure def decide(state: ConsensusState, p: Process): ConsensusState = {
+      val proposal = state.proposed.get(setChooseSome(state.proposed.keys().filter(x => state.proposed.get(x).size()>0)))
+      val decisions = state.decided.keys().filter(x => state.decided.get(x).size()>0)
+      val decision = if (decisions.size()>0) {state.decided.get(setChooseSome(decisions))}  else {proposal}
+      val newDecided = state.decided.set(p, decision)
+      {decided: newDecided, ...state}
+    }
+
+    // 4. state machine
+
+    var _consensus: ConsensusState
+
+    action consensusInit = all {
+        _consensus' = newConsensus(PROCESSES)
+    }
+
+    action consensusDoPropose = all {
+      nondet p = oneOf(PROCESSES)
+      nondet txs = oneOf(nonEmptyPowerset(TXS))
+      all {
+        require(_consensus.mayPropose(p, txs)),
+        _consensus' = _consensus.propose(p, txs)
+      }
+    }
+
+    action consensusDoDecide = all {
+      nondet p = oneOf(PROCESSES)
+      all {
+        require(_consensus.mayDecide(p)),
+        _consensus' = _consensus.decide(p)
+      }
+    }
+
+    action consensusStep = any {
+      consensusDoPropose,
+      consensusDoDecide
+    }
+
+    // 5. invariants
+        
+    // Validity: a process decides only proposed values and such values are non-empty set of transactions
+    pure def consensusValidityInv(state: ConsensusState): bool = {    
+        state.processes().forall(
+        p => or {
+    	       not(state.hasDecided(p)),
+    	       and {
+                   state.decisionOf(p).size()>0,
```

Comment:
Isn't this redundant with the other branch of the "or" clause? Is it related to non-short-circuit semantics?

Reply:
Indeed, thanks!